pydis_site/apps/api/serializers.py

```diff
@@ -249,15 +255,57 @@ class Meta:
         fields = ('id', 'name', 'colour', 'permissions', 'position')
 
 
-class UserSerializer(BulkSerializerMixin, ModelSerializer):
+class UserListSerializer(ListSerializer):
+    """List serializer for User model to handle bulk updates."""
+
+    def create(self, validated_data: list) -> list:
+        """Override create method to optimize django queries."""
+        present_users = User.objects.all()
+        new_users = []
+        present_user_ids = [user.id for user in present_users]
+
+        for user_dict in validated_data:
+            if user_dict["id"] in present_user_ids:
+                raise ValidationError({"id": "User already exists."})
```

Comment:
In fact, I'm pretty sure there's already a DB-level unique constraint (because the ID is the PK). This means we can rely on the constraint to fail and propagate an error.

Django's transaction control regarding `bulk_create` needs to be investigated. It says it "generally" uses one query, but what happens if it doesn't? Would it still do multiple queries in a single transaction? It's important because it needs to roll back _all_ created users if _any_ user fails to be created.

Reply:
The bulk_create() method actually can take a keyword argument called `ignore_conflicts`, if set to True, then it will ignore errors such as `ID already exists` error(basically, ignore duplicates) and only create new users.

Reply:
from docs:
`On databases that support it (all but Oracle), setting the ignore_conflicts parameter to True tells the database to ignore failure to insert any rows that fail constraints such as duplicate unique values. Enabling this parameter disables setting the primary key on each model instance (if the database normally supports it).`

Reply:
should I use this instead? what do you think?

Reply:
That's convenient, but no. I think it is best for the client to be notified of duplicates via an error response. Otherwise, the client may not realise that only a partial update occurred due to some being skipped (e.g. in the case there are duplicate users but the fields they modify are different).

Reply:
Oh, this is for create. Well, I'm a bit conflicted. On one hand, I don't like the idea of silently skipping things. On the other hand, it'd allow the syncer to sync as many users as it can rather than completely failing just cause 1 user is duplicated.

Therefore, I now think that if the user is in the DB already, then ignore duplicates. However, if the request data contains the user multiple times, then that's an error. Does that sound like a reasonable compromise?

Reply:
right, so if the number of duplicates is more than 1, then send an error message and also stop the creation process?
How about we ignore the duplicates and instead send a warning message, notifying the client that duplicates were found and skipped?
note: The only client we have now is the python bot, and I don't think there will be a problem with multiple duplicates.

Reply:
It's not about whether there are 1 or 2 or 3 duplicates. I'm suggesting it should depend on whether they are in the request or in the DB. Example: a request's data has two users with the ID 1. This should be an error. Another example: a requests data has one user with ID 1 and the DB has a user with ID 1. This should be skipped.

I don't know if this is worth the added complexity. I thought it'd make the API more robust. But it's also an option to just document that "duplicates will be skipped" and call it a day. Let's leave this as-is for now and see what anyone else has to say.